src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -2425,5 +2481,31 @@ public QueueItem PeekNextRequestForConnectionAttempt()
 
             public int RequestsWithoutAConnectionAttempt => _size - _attemptedConnectionsOffset;
         }
+
+        private sealed class HttpConnectionWaiter<T> : TaskCompletionSourceWithCancellation<T>
+        {
+            // When a connection attempt is pending, reference the connection's CTS, so we can tear it down if the initiating request is cancelled
+            // or completes on a different connection.
+            public volatile CancellationTokenSource? ConnectionCancellationTokenSource;
```

Comment:
This shouldn't need to be volatile if you just move the setting of it to null to be in the lock